machineconfiguration/v1/types_machineosconfig.go

```diff
@@ -0,0 +1,247 @@
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=machineosconfigs,scope=Cluster
+// +kubebuilder:subresource:status
+// +openshift:api-approved.openshift.io=https://github.com/openshift/api/pull/1773
+// +openshift:enable:FeatureGate=OnClusterBuild
+// +openshift:file-pattern=cvoRunLevel=0000_80,operatorName=machine-config,operatorOrdering=01
+// +kubebuilder:metadata:labels=openshift.io/operator-managed=
+
+// MachineOSConfig describes the configuration for a build process managed by the MCO
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=1
+type MachineOSConfig struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec describes the configuration of the machineosconfig
+	// +kubebuilder:validation:Required
+	Spec MachineOSConfigSpec `json:"spec"`
+
+	// status describes the status of the machineosconfig
+	// +optional
+	Status *MachineOSConfigStatus `json:"status,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineOSConfigList describes all configurations for image builds on the system
+//
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=1
+type MachineOSConfigList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	// items contains a collection of MachineOSConfig resources.
+	Items []MachineOSConfig `json:"items"`
+}
+
+// MachineOSConfigSpec describes user-configurable options as well as information about a build process.
+type MachineOSConfigSpec struct {
+	// machineConfigPool is the pool which the build is for
+	// +kubebuilder:validation:Required
+	MachineConfigPool MachineConfigPoolReference `json:"machineConfigPool"`
+	// buildInputs is where user input options for the build live
+	// +kubebuilder:validation:Required
+	BuildInputs BuildInputs `json:"buildInputs"`
+	// buildOutputs holds all information needed to handle booting the image after a build
+	// This currently contains a currentImagePullSecret field, which should be provided if the final pull secret used to pull the image to nodes from the registry
+	// is different than the one used for pushing the image to the registry during the build.
+	// +optional
+	BuildOutputs *BuildOutputs `json:"buildOutputs,omitempty"`
+}
+
+// MachineOSConfigStatus describes the status this config object and relates it to the builds associated with this MachineOSConfig
+type MachineOSConfigStatus struct {
```

Comment:
I'm surprised not to see conditions in this status?

Reply:
The alpha API had conditions but we removed them here since the status would be the same as on the build object. Do you think there's value in having the same conditions on both?

Reply:
Generally I'd expect to see conditions on most objects, and they would aggregate lower level objects.

For example, what would you do to report and issue in triggering the build? If there's no build object, how does the user know what went wrong